ports/zephyr/modzephyr.c

```diff
@@ -79,6 +92,238 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_0(mod_shell_net_iface_obj, mod_shell_net_iface);
 
 #endif // CONFIG_NET_SHELL
 
+#ifdef CONFIG_DISK_ACCESS
+STATIC const mp_obj_type_t zephyr_disk_type;
+
+typedef struct _zephyr_disk_obj_t {
+    mp_obj_base_t base;
+    const char *pdrv;
+    int block_size;
+} zephyr_disk_obj_t;
+
+STATIC void zephyr_disk_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    zephyr_disk_obj_t *self = self_in;
+    mp_printf(print, "%s", self->pdrv);
+}
+
+STATIC mp_obj_t zephyr_disk_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
+    mp_arg_check_num(n_args, n_kw, 1, 1, false);
+    zephyr_disk_obj_t *self = m_new_obj(zephyr_disk_obj_t);
+    self->base.type = type;
+    self->pdrv = mp_obj_str_get_str(args[0]);
+
+    if (disk_access_init(self->pdrv) != 0) {
+        mp_raise_ValueError("disk not found");
+    }
+
+    if (disk_access_ioctl(self->pdrv, DISK_IOCTL_GET_SECTOR_SIZE, &self->block_size)) {
+        mp_raise_ValueError("unable to get sector size");
+    }
+
+    return MP_OBJ_FROM_PTR(self);
+}
+
+STATIC mp_obj_t zephyr_disk_readblocks(mp_obj_t self_in, mp_obj_t block_num, mp_obj_t buf) {
+    zephyr_disk_obj_t *self = self_in;
+    mp_buffer_info_t bufinfo;
+    int ret;
+
+    mp_get_buffer_raise(buf, &bufinfo, MP_BUFFER_WRITE);
+    ret = disk_access_read(self->pdrv, bufinfo.buf, mp_obj_get_int(block_num), bufinfo.len / self->block_size);
+    return mp_obj_new_bool(ret == 0);
+}
+
+STATIC MP_DEFINE_CONST_FUN_OBJ_3(zephyr_disk_readblocks_obj, zephyr_disk_readblocks);
+
+STATIC mp_obj_t zephyr_disk_writeblocks(mp_obj_t self_in, mp_obj_t block_num, mp_obj_t buf) {
+    zephyr_disk_obj_t *self = self_in;
+    mp_buffer_info_t bufinfo;
+    int ret;
+
+    mp_get_buffer_raise(buf, &bufinfo, MP_BUFFER_READ);
+    ret = disk_access_write(self->pdrv, bufinfo.buf, mp_obj_get_int(block_num), bufinfo.len / self->block_size);
+    return mp_obj_new_bool(ret == 0);
+}
+
+STATIC MP_DEFINE_CONST_FUN_OBJ_3(zephyr_disk_writeblocks_obj, zephyr_disk_writeblocks);
+
+STATIC mp_obj_t zephyr_disk_ioctl(mp_obj_t self_in, mp_obj_t cmd_in, mp_obj_t arg_in) {
+    zephyr_disk_obj_t *self = self_in;
+    mp_int_t cmd = mp_obj_get_int(cmd_in);
+    int buf;
+
+    switch (cmd) {
+        case MP_BLOCKDEV_IOCTL_INIT:
+        case MP_BLOCKDEV_IOCTL_DEINIT:
+            return MP_OBJ_NEW_SMALL_INT(0);
+
+        case MP_BLOCKDEV_IOCTL_SYNC:
+            if (disk_access_ioctl(self->pdrv, DISK_IOCTL_CTRL_SYNC, &buf)) {
+                mp_raise_ValueError("unable to sync");
```

Comment:
Probably makes sense to return an error integer rather than raise here.  Maybe just return the int returned from this call?